src/plugins/discover/public/application/main/services/discover_state.ts

```diff
@@ -293,6 +312,8 @@ export function getState({
         stateStorage
       );
 
+      const { start, stop } = syncAppState();
```

Comment:
Calling `syncState` from within `initializeAndSync` fixes the back button state issue. This is because `syncState` takes a snapshot of the initial state to compare against before syncing state updates: https://github.com/elastic/kibana/blob/main/src/plugins/kibana_utils/public/state_sync/state_sync.ts#L151. When `syncState` is called from outside of `initializeAndSync`, the snapshot doesn't get reset when the data view is changed. Then when the user presses the back button, the new state appears to be the same as the initial state, so `syncState` ignores the update. Moving the call to `syncState` inside of `initializeAndSync` means the snapshot gets updated each time the data view changes and no state changes are ignored.